This patch implements unicode collation fallback for firmwares that set PlatformLang to local language (e.g. ru-RU),
yet do not support unicode collation prootocol or support it exclusively for en or en-US.
Discovered on Haswell GIGABYTE APTIO IV.

diff --git a/ShellPkg/Library/UefiShellCommandLib/UefiShellCommandLib.c b/ShellPkg/Library/UefiShellCommandLib/UefiShellCommandLib.c
index 0d37bd86..21b8e868 100644
--- a/ShellPkg/Library/UefiShellCommandLib/UefiShellCommandLib.c
+++ b/ShellPkg/Library/UefiShellCommandLib/UefiShellCommandLib.c
@@ -127,6 +127,39 @@ CommandInit(
         break;
       }
     }
+
+    for (Index = 0; Index < NumHandles && gUnicodeCollation == NULL && PlatformLang != NULL; Index++) {
+      //
+      // Open Unicode Collation Protocol
+      //
+      Status = gBS->OpenProtocol (
+                      Handles[Index],
+                      &gEfiUnicodeCollation2ProtocolGuid,
+                      (VOID **) &Uc,
+                      gImageHandle,
+                      NULL,
+                      EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                      );
+      if (EFI_ERROR (Status)) {
+        continue;
+      }
+
+      //
+      // Find any collation that supports en-US.
+      //
+      BestLanguage = GetBestLanguage (
+                       Uc->SupportedLanguages,
+                       FALSE,
+                       "en-US",
+                       NULL
+                       );
+      if (BestLanguage != NULL) {
+        FreePool (BestLanguage);
+        gUnicodeCollation = Uc;
+        break;
+      }
+    }
+
     if (Handles != NULL) {
       FreePool (Handles);
     }
